hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BlockSender.java

```diff
@@ -432,6 +432,7 @@
       ris = new ReplicaInputStreams(
           blockIn, checksumIn, volumeRef, fileIoProvider);
     } catch (IOException ioe) {
+      IOUtils.cleanupWithLogger(null, volumeRef);
```

Comment:
this is a good catch.

But even without exceptions, shouldn't we close the reference when the BlockSender is closed too?

Reply:
If there are no exceptions, the reference will be closed when its BlockSender is closed. I checked that the code that constructs the BlockSender closed it after it was used.